oss_client.cpp

```diff
@@ -0,0 +1,907 @@
+                                r->get_status(),
+                                r->get_hits(),
+                                req->m_keys - r->get_hits());
+
+            if (r->is_error()) {
+                benchmark_error_log("error response: %s\n", r->get_status());
+            }
+
+            handle_response(now, req, r);
+
+            sc->pending_resp--;
+            assert(sc->pending_reqs >= 0);
+
+            m_reqs_processed++;
+            responses_handled = true;
+        }
+        delete req;
+        if (error) {
+            return;
+        }
+    }
+
+    if (ret == -1) {
+        benchmark_error_log("error: response parsing failed.\n");
+    }
+
+    if (m_config->reconnect_interval > 0 && responses_handled) {
+        if ((m_reqs_processed % m_config->reconnect_interval) == 0) {
+            assert(m_pipeline->size() == 0);
+            benchmark_debug_log("reconnecting, m_reqs_processed = %u\n", m_reqs_processed);
+            disconnect();
+
+            ret = connect();
+            assert(ret == 0);
+
+            return;
+        }
+    }
+
+    fill_pipeline();
+}
+
+/*
+ * Utility function to get the object iterator type based on the config
+ */
+static inline
+int obj_iter_type(benchmark_config *cfg, unsigned char index)
+{
+    if (cfg->key_pattern[index] == 'R')
+        return OBJECT_GENERATOR_KEY_RANDOM;
+    else if (cfg->key_pattern[index] == 'G')
+        return OBJECT_GENERATOR_KEY_GAUSSIAN;
+    return OBJECT_GENERATOR_KEY_SET_ITER;
+}
+
+void oss_client::create_request(struct timeval timestamp)
+{
+    int cmd_size = 0;
+
+    // If the Set:Wait ratio is not 0, start off with WAITs
+    if (m_config->wait_ratio.b &&
+        (m_tot_wait_ops == 0 ||
+         (m_tot_set_ops/m_tot_wait_ops > m_config->wait_ratio.a/m_config->wait_ratio.b))) {
+
+        m_tot_wait_ops++;
+
+        unsigned int num_slaves = m_obj_gen->random_range(m_config->num_slaves.min, m_config->num_slaves.max);
+        unsigned int timeout = m_obj_gen->normal_distribution(m_config->wait_timeout.min,
+                                                              m_config->wait_timeout.max, 0,
+                                                              ((m_config->wait_timeout.max - m_config->wait_timeout.min)/2.0) + m_config->wait_timeout.min);
+
+        benchmark_debug_log("WAIT num_slaves=%u timeout=%u\n", num_slaves, timeout);
+
+        // currently wait commands goes to main connection
+        shard_connection* conn = MAIN_CONNECTION;
+
+        cmd_size = conn->protocol->write_command_wait(num_slaves, timeout);
+        conn->pipeline->push(new request(rt_wait, cmd_size, &timestamp, 0));
+        conn->pending_reqs++;
+        m_remain_reqs--;
+
+        if (conn->pipeline->size() >= m_config->pipeline) {
+            m_write_pipe = true;
+            m_full_pipe_id = 0;
+        }
+    }
+        // are we set or get? this depends on the ratio
+    else if (m_set_ratio_count < m_config->ratio.a) {
```

Comment:
Once again, a huge part of this is duplicated with client.c with no apparent reason.  It would make sense to pack it in a class hierarchy to share all identical behavior between the Redis Cluster and Standard Redis clients.